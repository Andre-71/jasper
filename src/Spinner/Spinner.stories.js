import React from "react"
import { storiesOf } from "@storybook/react"
import Spinner from "."

storiesOf("Spinner").add("Default", () => <Spinner />)
storiesOf("Spinner").add("Variant - Regular", () => (
  <Spinner variant="regular" />
))
storiesOf("Spinner").add("Size - Extra Small", () => <Spinner size="xs" />)
storiesOf("Spinner").add("Size - Small", () => <Spinner size="sm" />)
storiesOf("Spinner").add("Size - Medium", () => <Spinner size="md" />)
storiesOf("Spinner").add("Size - Large", () => <Spinner size="lg" />)
storiesOf("Spinner").add("Size - Extra Large", () => <Spinner size="xl" />)
storiesOf("Spinner").add("Speed - Slow", () => (
  <Spinner variant="regular" speed="slow" />
))
storiesOf("Spinner").add("Speed - Normal", () => (
  <Spinner variant="regular" speed="normal" />
))
storiesOf("Spinner").add("Speed - Fast", () => (
  <Spinner variant="regular" speed="fast" />
))
storiesOf("Spinner").add("Animation - Linear", () => (
  <Spinner variant="regular" animation="linear" />
))
storiesOf("Spinner").add("Animation - Ease", () => (
  <Spinner variant="regular" animation="ease" />
))
storiesOf("Spinner").add("Animation - Bouncy", () => (
  <Spinner variant="regular" animation="bouncy" />
))
