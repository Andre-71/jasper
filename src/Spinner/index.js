import React from "react"

import PropTypes from "prop-types"
import styles from "./styles.module.css"

const Spinner = (props) => {
  const { variant, size, speed, animation, ...other } = props
  return (
    <div className={styles.spinner}>
      <div
        className={`${styles[variant]} ${styles[size]} ${styles[speed]} ${styles[animation]}`}
        {...other}
      />
    </div>
  )
}

Spinner.propTypes = {
  variant: PropTypes.string,
  size: PropTypes.string,
  speed: PropTypes.string,
  animation: PropTypes.string,
}
Spinner.defaultProps = {
  variant: "default",
  size: "md",
  speed: "normal",
  animation: "linear",
}

export default Spinner
