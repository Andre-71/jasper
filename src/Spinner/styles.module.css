.spinner {
  display: flex;
  justify-content: center;
  align-items: center;
  min-height: 98vh;
}

.default {
  -webkit-animation: spin var(--speed) var(--animation-type) infinite;
  animation: spin var(--speed) var(--animation-type) infinite;
  border-radius: 50%;
  height: var(--size);
  width: var(--size);
  position: relative;
}

.default:before,
.default:after {
  content: "";
  position: absolute;
}

.default:before {
  border-radius: 50%;
  background: linear-gradient(45deg, darkorange, mediumseagreen) 0% 0%,
    linear-gradient(90deg, mediumseagreen, blue) 100% 0%,
    linear-gradient(200deg, blue, red) 100% 100%,
    linear-gradient(360deg, red, darkorange) 0% 100%;
  background-repeat: no-repeat;
  background-size: 50% 50%;
  top: -1px;
  bottom: -1px;
  left: -1px;
  right: -1px;
}

.default:after {
  background: white;
  border-radius: 50%;
  top: 10%;
  bottom: 10%;
  left: 10%;
  right: 10%;
}

.regular {
  -webkit-animation: spin var(--speed) var(--animation-type) infinite;
  animation: spin var(--speed) var(--animation-type) infinite;
  border: var(--thickness) solid #f3f3f3;
  border-radius: 50%;
  border-top: var(--thickness) solid black;
  width: var(--size);
  height: var(--size);
}

.xs {
  --size: 30px;
  --thickness: 4px;
}

.sm {
  --size: 60px;
  --thickness: 8px;
}

.md {
  --size: 120px;
  --thickness: 16px;
}

.lg {
  --size: 240px;
  --thickness: 32px;
}

.xl {
  --size: 480px;
  --thickness: 64px;
}

.slow {
  --speed: 1.5s;
}

.normal {
  --speed: 1s;
}

.fast {
  --speed: 0.5s;
}

.linear {
  --animation-type: linear;
}

.ease {
  --animation-type: ease-in-out;
}

.bouncy {
  --animation-type: cubic-bezier(0.26, 1.36, 0.74, 0.92);
}

@-webkit-keyframes spin {
  0% {
    -webkit-transform: rotate(0deg);
  }
  100% {
    -webkit-transform: rotate(360deg);
  }
}

@keyframes spin {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
