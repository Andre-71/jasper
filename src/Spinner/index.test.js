import React from 'react'
import Spinner from '.'
import renderer from 'react-test-renderer'

it('renders correctly', () => {
  let tree = renderer
    .create(<Spinner />)
    .toJSON()
  expect(tree).toMatchSnapshot()
  
  tree = renderer
    .create(<Spinner variant="regular" />)
    .toJSON()
  expect(tree).toMatchSnapshot()
  
  tree = renderer
    .create(<Spinner size="lg" />)
    .toJSON()
  expect(tree).toMatchSnapshot()

  tree = renderer
    .create(<Spinner speed="fast" />)
    .toJSON()
  expect(tree).toMatchSnapshot()

  tree = renderer
    .create(<Spinner animation="bouncy" />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})