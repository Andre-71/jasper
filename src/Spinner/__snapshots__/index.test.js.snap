// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="spinner"
>
  <div
    className="default md normal linear"
  />
</div>
`;

exports[`renders correctly 2`] = `
<div
  className="spinner"
>
  <div
    className="regular md normal linear"
  />
</div>
`;

exports[`renders correctly 3`] = `
<div
  className="spinner"
>
  <div
    className="default lg normal linear"
  />
</div>
`;

exports[`renders correctly 4`] = `
<div
  className="spinner"
>
  <div
    className="default md fast linear"
  />
</div>
`;

exports[`renders correctly 5`] = `
<div
  className="spinner"
>
  <div
    className="default md normal bouncy"
  />
</div>
`;
